pkg/alertmanager/distributor/distributor.go

```diff
@@ -0,0 +1,306 @@
+	}
+}
+
+func (d *Distributor) doWrite(userID string, w http.ResponseWriter, r *http.Request) {
+	source := util.GetSourceIPsFromOutgoingCtx(r.Context())
+
+	var body []byte
+	var err error
+	if r.Body != nil {
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			level.Error(d.logger).Log("msg", "failed to read the request body during write", "err", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	err = ring.DoBatch(r.Context(), RingOp, d.alertmanagerRing, []uint32{shardByUser(userID)}, func(am ring.InstanceDesc, _ []int) error {
+		d.amSends.WithLabelValues(am.Addr).Inc()
+
+		// Use a background context to make sure all alertmanagers get the request even if we return early.
+		localCtx, cancel := context.WithTimeout(context.Background(), d.cfg.RemoteTimeout)
+		defer cancel()
+		localCtx = user.InjectOrgID(localCtx, userID)
+		if sp := opentracing.SpanFromContext(r.Context()); sp != nil {
+			localCtx = opentracing.ContextWithSpan(localCtx, sp)
+		}
+		// Get clientIP(s) from Context and add it to localCtx
+		localCtx = util.AddSourceIPsToOutgoingContext(localCtx, source)
+
+		resp, err := d.doRequest(localCtx, am, r, body)
+		if err != nil {
+			d.amSendFailures.WithLabelValues(am.Addr).Inc()
+			return err
+		}
+
+		if resp.GetStatus() != alertmanagerpb.OK {
+			return errors.New("alertmanager grpc request not ok")
+		}
+
+		return nil
+	}, func() {})
+
+	if err == nil {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+
+	d.respondFromError(err, w)
+}
+
+func (d *Distributor) doRead(userID string, w http.ResponseWriter, r *http.Request) {
+	key := shardByUser(userID)
+	replicationSet, err := d.alertmanagerRing.Get(key, RingOp, nil, nil, nil)
+	if err != nil {
+		level.Error(d.logger).Log("msg", "failed to get replication set from the ring", "err", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Until we have a mechanism to combine the results from multiple alertmanagers,
+	// we forward the request to only only of the alertmanagers.
+	amDesc := replicationSet.Ingesters[0]
+
+	var body []byte
+	if r.Body != nil {
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			level.Error(d.logger).Log("msg", "failed to read the request body during reads", "err", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	d.amSends.WithLabelValues(amDesc.Addr).Inc()
+	resp, err := d.doRequest(r.Context(), amDesc, r, body)
+	if err != nil {
+		d.amSendFailures.WithLabelValues(amDesc.Addr).Inc()
+		d.respondFromError(err, w)
+		return
+	}
+
+	if resp.GetStatus() != alertmanagerpb.OK {
+		http.Error(w, resp.Error, http.StatusInternalServerError)
+		return
+	}
+
+	http.Error(w, string(resp.HttpResponse.Body), int(resp.HttpResponse.Code))
+}
+
+func (d *Distributor) respondFromError(err error, w http.ResponseWriter) {
+	httpResp, ok := httpgrpc.HTTPResponseFromError(errors.Cause(err))
+	if !ok {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```

Comment:
I don't think the user needs to know about this error, let's log and use a generic message something like: `Failed to process the request to the alertmanager`